### PR TITLE
Add method for parsing forms to request object

### LINKED ISSUE #347: Automatically decode x-www-form-urlencoded and multipart/form-data HTTP request bodies
Make it easy to parse data posted directly from HTML forms to an HTTP triggered function. This may be particularly useful for Azure Static Web Apps but in general this comes up regularly for Functions.

Examples from recent blog posts:
* @svdoever https://www.sergevandenoever.nl/Processing-multipart-form-data-in-nodejs-azure-function-with-httptrigger/
* @estruyf https://www.eliostruyf.com/parse-multipart-form-data-azure-function/

```html
<h1>x-www-form-urlencoded</h1>
<form action="http://localhost:7071/api/HttpTrigger1" method="POST">
  <div>
    Name: <input name="name" value="Azure Functions" />
  </div>
  <div>
    Greeting: <input name="greeting" value="Hello" />
  </div>
  <div>
    <input type="submit" value="Submit" />
  </div>
</form>
```

```html
<h1>multipart/form-data</h1>
<form action="http://localhost:7071/api/HttpTrigger1" method="POST" enctype="multipart/form-data">
  <div>
    Name: <input name="name" value="Azure Functions" />
  </div>
  <div>
    Greeting: <input name="greeting" value="Hello" />
  </div>
  <div>
    <input type="file" name="myfile">
  </div>
  <div>
    <input type="submit" value="Submit" />
  </div>
</form>
```

Thinking something like:

```js
module.exports = async function (context, req) {
    const form = req.formPostData();
    const { name, greeting, myfile } = form;

    // myfile:
    // {
    //     "content-type": "image/jpeg",
    //     "data": <buffer>
    // }

    const fileContentBuffer = myfile.data;
    context.log(`name: ${name} - greeting: ${greeting}`);
    context.log(`filename: ${myfile.filename} - content type: ${myfile['content-type']} - bytes: ${fileContentBuffer.length}`);
    context.res = {
        status: 302,
        headers: {
            "location": "https://microsoft.com/thankyou.html"
        }
    };
}
```

Related: #146 

/cc @mhoeger @johnpapa @craigshoemaker @burkeholland

### PR DESCRIPTION
**Pull request by ejizba**
See https://github.com/Azure/azure-functions-nodejs-worker/issues/347#issue-740212555 for good examples from Anthony

See https://github.com/Azure/azure-functions-nodejs-worker/issues/347#issuecomment-984036447 for my description of the API proposal

Fixes https://github.com/Azure/azure-functions-nodejs-worker/issues/347

### CODE REVIEW

**Comment by ejizba** [comment]
ping @alrod @hossam-nasr

File: src/parsers/parseMultipartForm.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+import { FormPart } from '@azure/functions';
+import { HeaderName } from '../constants';
+import { getHeaderValue, parseContentDisposition } from './parseHeader';
+
+const carriageReturn = Buffer.from('\r')[0];
+const newline = Buffer.from('\n')[0];
+
+// multipart/form-data specification https://datatracker.ietf.org/doc/html/rfc7578
+export async function parseMultipartForm(chunk: Buffer, boundary: string): Promise<[string, FormPart][]> {
```
Comment (alrod) [inline src/parsers/parseMultipartForm.ts]: Why the function is `async`? it looks like we did not call `await` inside.

File: src/parsers/parseMultipartForm.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                } else {
+                    partValueStart = lineStart;
+                    partValueEnd = lineStart;
+                    result.push([currentName, currentPart]);
+                }
+            } else {
+                const contentDisposition = getHeaderValue(lineAsString, HeaderName.contentDisposition);
+                if (contentDisposition) {
+                    const [, dispositionParts] = parseContentDisposition(contentDisposition);
+                    currentName = dispositionParts.get('name');
```
Comment (alrod) [inline src/parsers/parseMultipartForm.ts:73]: Are `'name'` and `'fileName'` always in camel case?

File: test/parseForm.test.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { expect } from 'chai';
+import 'mocha';
+import { parseForm } from '../src/parsers/parseForm';
+
+describe('parseForm', () => {
+    describe('multipart', () => {
+        const contentType = 'multipart/form-data; boundary=----WebKitFormBoundaryeJGMO2YP65ZZXRmv';
```
Comment (alrod) [inline test/parseForm.test.ts:10]: Should we define a const for the string "`WebKitFormBoundaryeJGMO2YP65ZZXRmv`" ? It's repeated multiple times.

File: src/parsers/parseForm.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            const parsed = new URLSearchParams(data.toString());
+            const parts: [string, types.FormPart][] = [];
+            for (const [key, value] of parsed) {
+                parts.push([key, { value: Buffer.from(value) }]);
+            }
+            return new Form(parts);
+        }
+        default:
+            throw new Error(
+                `Media type "${mediaType}" does not match types supported for form parsing: "${MediaType.multipartForm}", "${MediaType.urlEncodedForm}".`
```
Comment (alrod) [inline src/parsers/parseForm.ts:30]: Do we have a test for the error case?

**Review by alrod — DISMISSED** [review]

File: src/parsers/parseMultipartForm.ts (hunk 1 above)
Comment (ejizba) [reply src/parsers/parseMultipartForm.ts]: Removed async

File: src/parsers/parseMultipartForm.ts (hunk 2 above)
Comment (ejizba) [reply src/parsers/parseMultipartForm.ts:73]: The header names are case insensitive, so it shouldn't matter. The `HeaderParams` class already handles this and I have related tests in `parseHeader.test.ts`

File: test/parseForm.test.ts (hunk 3 above)
Comment (ejizba) [reply test/parseForm.test.ts:10]: If it was for production code I would say yes, but for test code I often prefer to hard-code the test data even if there's duplicates. It just feels closer to the real deal when the inputs/outputs of the test are simple values without extra manipulation. Sometimes (but not always) I find it more readable this way, too.

File: src/parsers/parseForm.ts (hunk 4 above)
Comment (ejizba) [reply src/parsers/parseForm.ts:30]: Added some error test cases

**Comment by ejizba** [comment]
ping @alrod

**Review by alrod — APPROVED** [review]

**Pull request merged**